File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> answer;
    void dfs(int pos, vector<int>& candidates, int target, vector<int>& path){
        if(target == 0){
            answer.push_back(path);
            return;
        }
        if(target < 0 || pos >= candidates.size()){
            return;
        }
        for(int i = pos; i < candidates.size(); i++){
            if(i > pos && candidates[i] == candidates[i- 1]){
                continue;
            }
            path.push_back(candidates[i]);
            dfs(i + 1, candidates, target - candidates[i], path);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> path;
        sort(candidates.begin(), candidates.end());
        dfs(0, candidates, target, path);
        return answer;
    }
};

File: 40-combination-sum-ii/40-combination-sum-ii.cs
public class Solution {
    IList<IList<int>> answer = new List<IList<int>>();
    public void dfs(int pos, int[] candidates, int target, IList<int> path)
    {
        if(target == 0)
        {
            answer.Add(path.ToList());
            return;
        }
        if(target < 0 || pos >= candidates.Length)
        {
            return;
        }
        for(int i = pos; i < candidates.Length; i++){
            if(i > pos && candidates[i] == candidates[i - 1])
            {
                continue;
            }
            path.Add(candidates[i]);
            dfs(i + 1, candidates, target - candidates[i], path);
            path.RemoveAt(path.Count - 1);
        }
    }
    public IList<IList<int>> CombinationSum2(int[] candidates, int target) {
        IList<int> path = new List<int>();
        Array.Sort(candidates);
        dfs(0, candidates, target, path);
        return answer;
    }
}